### ISSUE TITLE
Have multiple passes by SemanticAnalyzerPass3 to resolve more forward references.

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#4095]
As mentioned in https://github.com/python/mypy/pull/3952#issuecomment-330804438 some double-forward references (and other corner cases) are currently not resolved. For example this does not work:
```python
x: A
A = B
B = List[int]
```
This is because third pass happens only once. In ``checker.py`` we have deferred nodes, and there are mini-passes over them until they are all done, I think we can adopt the same idea here and have repeated third passes.

This is however low priority since these edge cases are extremely rare and don't cause crashes, only false positives.

**Comment by ilevkivskyi** [comment]
Related #731

**Comment by gvanrossum** [comment]
Also that particular example won't work -- even if you protect the `x: A` from crashing (e.g. using `x: 'A'`), `A = B` will still crash.

**Comment by ilevkivskyi** [comment]
Yes, this is just a simple illustration, there are other more complex situations that will work at runtime, but not in mypy, all of them are purely "academic" however.

**Comment by ilevkivskyi** [comment]
Even this one formally speaking can be protected by ``if TYPE_CHECKING: ...``

**Comment by ilevkivskyi** [comment]
This now works correctly with the new analyzer.
